Mathlib/Topology/Compactness/Lindelof.lean

```diff
@@ -0,0 +1,357 @@
+there exists a countable subfamily whose intersection avoids `s`. -/
+theorem isLindelof_of_countable_subfamily_closed
+    (h : ∀ {ι : Type u} (t : ι → Set X), (∀ i, IsClosed (t i)) → (s ∩ ⋂ i, t i) = ∅ →
+    ∃ u : Set ι, u.Countable ∧ (s ∩ ⋂ i ∈ u, t i) = ∅) :
+    IsLindelof s :=
+  isLindelof_of_countable_subcover fun U hUo hsU ↦ by
+    rw [← disjoint_compl_right_iff_subset, compl_iUnion, disjoint_iff] at hsU
+    rcases h (fun i ↦ (U i)ᶜ) (fun i ↦ (hUo _).isClosed_compl) hsU with ⟨t, ht⟩
+    refine ⟨t, ?_⟩
+    rwa [← disjoint_compl_right_iff_subset, compl_iUnion₂, disjoint_iff]
+
+/-- A set `s` is Lindelöf if and only if
+for every open cover of `s`, there exists a countable subcover. -/
+theorem isLindelof_iff_countable_subcover :
+    IsLindelof s ↔ ∀ {ι : Type u} (U : ι → Set X),
+      (∀ i, IsOpen (U i)) → (s ⊆ ⋃ i, U i) → ∃ t : Set ι, t.Countable ∧ s ⊆ ⋃ i ∈ t, U i :=
+  ⟨fun hs ↦ hs.elim_countable_subcover, isLindelof_of_countable_subcover⟩
+
+/-- A set `s` is Lindelöf if and only if
+for every family of closed sets whose intersection avoids `s`,
+there exists a countable subfamily whose intersection avoids `s`. -/
+theorem isLindelof_iff_countable_subfamily_closed :
+    IsLindelof s ↔ ∀ {ι : Type u} (t : ι → Set X),
+    (∀ i, IsClosed (t i)) → (s ∩ ⋂ i, t i) = ∅
+    → ∃ u : Set ι, u.Countable ∧ (s ∩ ⋂ i ∈ u, t i) = ∅ :=
+  ⟨fun hs ↦ hs.elim_countable_subfamily_closed, isLindelof_of_countable_subfamily_closed⟩
+
+/-- The empty set is a Lindelof set. -/
+@[simp]
+theorem isLindelof_empty : IsLindelof (∅ : Set X) := fun _f hnf _ hsf↦
+  Not.elim hnf.ne <| empty_mem_iff_bot.1 <| le_principal_iff.1 hsf
+
+/-- A singleton set is a Lindelof set. -/
+@[simp]
+theorem isLindelof_singleton {x : X} : IsLindelof ({x} : Set X) := fun f hf _ hfa ↦
+  ⟨x, rfl, ClusterPt.of_le_nhds'
+    (hfa.trans <| by simpa only [principal_singleton] using pure_le_nhds x) hf⟩
+
+theorem Set.Subsingleton.isLindelof (hs : s.Subsingleton) : IsLindelof s :=
+  Subsingleton.induction_on hs isLindelof_empty fun _ ↦ isLindelof_singleton
+
+/-- X is a Lindelöf space iff every open cover has a countable subcover.-/
+class LindelofSpace (X : Type*) [TopologicalSpace X] : Prop where
+  /-- In a Lindelöf space, `Set.univ` is a Lindelöf set. -/
+  isLindelof_univ : IsLindelof (univ : Set X)
+
+theorem isLindelof_univ_iff : IsLindelof (univ : Set X) ↔ LindelofSpace X :=
+  ⟨fun h => ⟨h⟩, fun h => h.1⟩
+
+theorem isLindelof_univ [h : LindelofSpace X] : IsLindelof (univ : Set X) :=
+  h.isLindelof_univ
+
+/-- A compact set `s` is Lindelöf. -/
+theorem IsCompact.isLindelof (hs : IsCompact s) :
+    IsLindelof s := by tauto
+
+/-- A compact space `X` is Lindelöf. SHOULD THIS BE AN INSTANCE? -/
+instance [CompactSpace X] : LindelofSpace X := { isLindelof_univ := isCompact_univ.isLindelof}
+
+/-- `X` is a non-Lindelöf topological space if it is not a Lindelöf space. -/
+class NonLindelofSpace (X : Type*) [TopologicalSpace X] : Prop where
```

Comment:
Do you plan to add any instances of this typeclass? Any theorems about it?

Reply:
Yes, but I wanted to keep the PR small! I have about twice as much code about Lindelöf spaces lying arounds, including some basic results about it. Do you want me to add them to this PR already?